### PR TITLE
feat: make ConversationEventProcessor a ZMEventAsyncConsumer - WPB-5138

### PR DESCRIPTION
**Pull request by netbe**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5138" title="WPB-5138" target="_blank">TaskWPB-5138</a>  [iOS] Adapt remove member to be async
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Continuing the work on #635, ConversationEventProcessor need to become async,  since it will contain calls to MLSService methods.

### Solutions

* Introduce a new ZMEventAsyncConsumer so we don't touch other consumers
* Use ConversationEventProcessor as eventAsyncConsumer in the EventProcessor.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/Synchronization/EventProcessor.swift [hunk 1]
```diff
@@ -148,14 +151,20 @@ actor EventProcessor: UpdateEventProcessor {
...
-                for event in decryptedUpdateEvents {
+            for event in decryptedUpdateEvents {
+                await syncContext.perform {
                     for eventConsumer in self.eventConsumers {
-                        // TODO: [jacob] EventConsumer.processEvents should become async
                         eventConsumer.processEvents([event], liveEvents: true, prefetchResult: prefetchResult)
                     }
-                    self.eventProcessingTracker.registerEventProcessed()
                 }
+                // TODO: [F] @Jacob should this be done on syncContext to keep every thing in sync?
```
Comment (typfel) [inline wire-ios-sync-engine/Source/Synchronization/EventProcessor.swift:169]: We can't call an async function inside a `syncContext.perform {}`, but Isn't this already in sync? We are awaiting each call to `processEvents()`

Comment (netbe) [reply wire-ios-sync-engine/Source/Synchronization/EventProcessor.swift:169]: exactly, I thought this was not needed.
but I remember you mention something about respecting that other eventConsumers are supposed to run on syncContext

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationRequestStrategy.swift [hunk 2]
```diff
@@ -373,18 +362,6 @@ extension ConversationRequestStrategy: ZMUpstreamTranscoder {
     public func updateUpdatedObject(_ managedObject: ZMManagedObject,
                                     requestUserInfo: [AnyHashable: Any]? = nil,
                                     response: ZMTransportResponse, keysToParse: Set<String>) -> Bool {
-
-        guard
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationRequestStrategy.swift]: @typfel can you confirm we don't need to parse the payloads here with ConversationEventProcessor?

Comment (typfel) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationRequestStrategy.swift]: We still need to parse the payloads here with `ConversationEventProcessor`. This was accidentally removed when replacing the `ZMEventConsumer` with `ZMAsyncEventConsumer`?

File: wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift [hunk 3]
```diff
@@ -195,10 +195,12 @@ extension ConnectionRequestStrategy: ZMEventConsumer {
             case .userConnection:
                 if let conversationEvent = Payload.UserConnectionEvent(payloadData) {
                     let processor = ConnectionPayloadProcessor()
-                    processor.processPayload(
-                        conversationEvent,
-                        in: managedObjectContext
-                    )
+                    managedObjectContext.performAndWait {
```
Comment (typfel) [inline wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift]: why is this `performAndWait` necessary?  The non-async`processEvents` is should always be executed on the sync context?

**Review by typfel — CHANGES_REQUESTED** [review]

File: wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift (hunk 3 above)
Comment (netbe) [reply wire-ios-request-strategy/Sources/Request Strategies/Connection/ConnectionRequestStrategy.swift]: I guess it's a left over of what I tried before, I'll remove

File: wire-ios-sync-engine/Source/Synchronization/EventProcessor.swift (hunk 1 above)
Comment (typfel) [reply wire-ios-sync-engine/Source/Synchronization/EventProcessor.swift:169]: Don't remember I think it's fine like this.

**Review by typfel — APPROVED** [review]

**Review by caldrian — APPROVED** [review]

**Pull request merged**